### ISSUE TITLE
Improve the robustness of connections by session reuse

### ISSUE THREAD

**Issue by promi** [mumble-voip/mumble#3972]
We are running a Mumble server that constantly has 20 to 40 uses connected. What we're seeing is that some users have periods where they constantly disconnect and reconnect.

That is not optimal for several reasons:

One reason is that the other users get spammed with disconnect and connect messages (that will be partially solved by the chat renovation #2521, for which I had an alternate proposal #3941).

The user also causes status updates in that channel he is in, so the channel list constantly updates.

For users with shaky connections (like WIFI or mobile) the experience is also not great they need to wait for mumble to reconnect.

Other applications like Skype or TeamSpeak or Discord handle this in a much better way than Mumble currently does.

I propose that the handling should be changed in a way that users are not constantly disconnected and reconnected any longer. Simply increasing the timeout is probably not enough. There should be a way that a client can reconnect with a fresh connection and reuse the old session.

Is that a feasable approach or does anyone have a better idea?

**Comment by Krzmbrzl** [comment]
Uhm I don't know if that is possible.

First off: I guess there's nothing we can do about people having a disconnect due to a bad internet connection in the first place. Therefore this can only be about handling the re-connect.

Now I don't know the details of how this is currently implemented but from my point of view the following had to be done for this to work:
1. The server has to somehow know that this client has been connected just a moment before already and it also has to remmeber the session ID of that user.
I guess the recognizing part could be done via the client's certificate and storing a certificate-ID map should be doable. The question to be asked then is how long may a user be disconnected before this data gets deleted? That's a minor detail for now though
2. The clients have to remember the disconnected client for the same amount of time.

Point 2 is the thing that I think will render this impossible. If a user disconnects, we want Mumble's UI to reflect that change ASAP. We don't want to have ghost-users hanging around in our UI for minutes, do we?
Thus the client will remove the disconnected user immediately. Thus even if the server remembered the disconnected client on a re-connect and assigns it its original ID and puts it in the original channel, it would still have to send an update to the clients for them to know that this client is back again.

To me this sounds as if it would be quite some work for something that doesn't really differ much from what we currently have.
Sure the messages would be gone but as you already mentioned this problem is being dealt with in different ways already.

The only way I could see this work would be by introducing a new client state "lagging" (or something like this) indicating that the client was disconnected because of a timeout. Then for these clients we'd need to explicitly turn them into ghost-users in the UI and do the rest as mentioned above.
However we'd still need to send an update to the clients once that user is back setting their state to connected again.

So yet again I'm not sure if much is gained here...

**Comment by promi** [comment]
I like the idea of this lagging state and then ghosting the user. After a certain amount of time the user would then go offline and disappear completely. 

One additional gain is that you can see why the user has left the server. In the current state it is impossible to tell if someone leaves because of timeout or because he intentionally left.

From the perspective of the lagging user Mumble could also give some visual clue that the user is lagging. He would still see the tree in the state it was since he lost the connection.

This would also be a foundation for sending offline messages which could be added later.

**Comment by Krzmbrzl** [comment]
In my opinion though such a ghosting user has to be marked so it is clear that the client belonging to that ghosting user is no longer connected.

Technically speaking though I don't think that it is necessary for the server to reuse the session ID. I think it should be enough to store in which channel that user has been. And that could be achieved by letting that user ghost around on the server as well. This would also turn the ghost-handling into something that is handled server-sided instead of client-sided.

> He would still see the tree in the state it was since he lost the connection.

Nah I don't think that's a good idea... Could lead to confusion where people think they are still connected but aren't...

> This would also be a foundation for sending offline messages which could be added later.

I don't really see a connection to offline messaging here

**Comment by sudoforge** [comment]
I haven't seen the reported issue addressed yet:

> One reason is that the other users get spammed with disconnect and connect messages
> ...
> The user also causes status updates in that channel he is in, so the channel list constantly updates.

I'm not able to verify it at this moment, but I'm fairly certain that both the auditory and text notifications can be disabled in the settings window.

**Comment by promi** [comment]
> I'm not able to verify it at this moment, but I'm fairly certain that both the auditory and text notifications can be disabled in the settings window.

You're right about that.

I do that as a workaround, but then I see nobody connect and disconnect any longer. That's not a real solution to me.

**Comment by sudoforge** [comment]
> I do that as a workaround, but then I see nobody connect and disconnect any longer. That's not a real solution to me.

I agree that it's not the best "solution"; what you've proposed would be a great alternative, that I think would improve the user experience.
